ocaml/xapi/xapi_vdi.ml

```diff
@@ -758,4 +758,87 @@ let read_database_pool_uuid ~__context ~self =
 	| Some (_, uuid) -> uuid
 	| None -> ""
 
+module VDI_CStruct = struct
+
+	let magic_number = 0x7ada7adal
+	let magic_number_offset = 0
+	let version = 1l
+	let version_offset = 4
+	let length_offset = 8
+	let data_offset = 12
+	let vdi_format_length = 12 (* VDI format takes 12bytes *)
+	let vdi_size = 4194304 (* 4MiB *)
+	let default_offset = 0
+
+	(* Set the magic number *)
+	let set_magic_number cstruct =
+		Cstruct.BE.set_uint32 cstruct magic_number_offset magic_number
+
+	(* Get the magic number *)
+	let get_magic_number cstruct =
+		Cstruct.BE.get_uint32 cstruct magic_number_offset
+
+	(* Set the version *)
+	let set_version cstruct =
+		Cstruct.BE.set_uint32 cstruct version_offset version
+
+	(* Set the data length *)
+	let set_data_length cstruct len =
+		Cstruct.BE.set_uint32 cstruct length_offset len
+
+	(* Get the data length *)
+	let get_data_length cstruct =
+		Cstruct.BE.get_uint32 cstruct length_offset
+
+	(* Write the string to the vdi *)
+	let write_to_vdi cstruct text text_len =
+		Cstruct.blit_from_string text default_offset cstruct data_offset text_len;
+		set_data_length cstruct (Int32.of_int text_len)
+
+	(* Read the string from the vdi *)
+	let read_from_vdi cstruct =
+		let curr_len = Int32.to_int (get_data_length cstruct) in
+		let curr_text = String.make curr_len '\000' in
+		Cstruct.blit_to_string cstruct data_offset curr_text default_offset curr_len;
+		curr_text
+
+	(* Format the vdi for the first time *)
+	let format_rrd_vdi cstruct =
+		set_magic_number cstruct;
+		set_version cstruct
+
+end
+
+let op_on_device device =
```

Comment:
This function is a bit confusingly named, as it doesn't actually do an operation on the device but just returns its contents as a cstruct.

What would be nice is to have a function of type

```
val op_on_device : __context:Context.t -> vdi:API.ref_VDI -> (Cstruct.t -> 'a) -> 'a
```

This would

- attach the disk to dom0 (using `Sm_fs_ops.with_block_attached_device`)
- map the resulting device as a `Cstruct.t`
- pass the `Cstruct.t` to the supplied function
- return the result of the function

Also, I see you're passing `O_RDWR` to `Unix.openfile` unconditionally - not sure what would happen here if tapdisk is in read-only mode. It would be better to use the same read/write flags as is passed to `Sm_fs_ops.with_block_attached_device`.